cpp/src/arrow/io/buffered_test.cc

```diff
@@ -491,6 +491,30 @@ TEST_F(TestBufferedInputStream, BufferSizeLimit) {
   }
 }
 
+TEST_F(TestBufferedInputStream, PeekPastBufferedBytes) {
+  // GH-43949: Peek and SetBufferSize should not affect the
+  // buffered bytes.
+  MakeExample1(/*buffer_size=*/10, default_memory_pool(), /*raw_read_bound=*/15);
+  ASSERT_OK_AND_ASSIGN(auto bytes, buffered_->Read(9));
+  EXPECT_EQ(std::string_view(bytes->data_as<char>(), bytes->size()),
+            kExample1.substr(0, 9));
```

Comment:
Normally this should work (`Buffer` has a conversion operator to `std::string_view`):
```suggestion
  EXPECT_EQ(std::string_view(*bytes), kExample1.substr(0, 9));
```